File: 547-NumberOfProvinces/547-NumberOfProvinces.py
# Last updated: 7/24/2025, 8:30:28 AM
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        
        n = len(isConnected)
        notVisited = [True]*n

        def helperDFS(node: int):
            notVisited[node] = False
            for i in range(n):
                if notVisited[i] and isConnected[node][i]:
                    helperDFS(i)
            
        ans = 0
        for i in range(n):
            if notVisited[i]:
                ans += 1
                helperDFS(i)
        return ans
        

File: 547-NumberOfProvinces/547-NumberOfProvinces.cpp
// Last updated: 7/24/2025, 8:30:22 AM
class Solution {
private:
    int res = 0;
    vector<int> parent;

    int find(int i) {
        if(parent[i] != i)
            parent[i] = find(parent[i]);
        return parent[i];
    }

    void make(int n) {
        parent.resize(n);
        for(int i = 0; i < n; i++)
            parent[i] = i;
    }
    
    void connect(int a, int b) {
        int parentA = find(a);
        int parentB = find(b); 

        if(parentA == parentB)
            return;
        
        parent[parentA] = parentB;
    }

public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        make(n);

        for(int i = 0; i < n; i++)
            for(int j = 0; j < n; j++)
                if(isConnected[i][j])
                    connect(i, j);
        
        for(int i = 0; i < n; i++)
            if(parent[i] == i)
                res++;

        return res;
    }
};